src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java

```diff
@@ -470,45 +464,24 @@ public Object run() {
             try {
                 m = javaAgentClass.getDeclaredMethod(methodname,
                                                  new Class<?>[] { String.class });
-            } catch (NoSuchMethodException x) {
-                // ignore this exception because we'll try
-                // two arg inheritance next
-            }
-        }
-
-        if (m == null) {
-            // now try the inherited 2-arg method
-            try {
-                m = javaAgentClass.getMethod( methodname,
-                                 new Class<?>[] {
-                                     String.class,
-                                     java.lang.instrument.Instrumentation.class
-                                 }
-                               );
-                twoArgAgent = true;
-            } catch (NoSuchMethodException x) {
-                // ignore this exception because we'll try
-                // one arg inheritance next
-            }
-        }
-
-        if (m == null) {
-            // finally try the inherited 1-arg method
-            try {
-                m = javaAgentClass.getMethod(methodname,
-                                             new Class<?>[] { String.class });
             } catch (NoSuchMethodException x) {
                 // none of the methods exists so we throw the
                 // first NoSuchMethodException as per 5.0
                 throw firstExc;
             }
         }
 
-        // the premain method should not be required to be public,
-        // make it accessible so we can call it
-        // Note: The spec says the following:
-        //     The agent class must implement a public static premain method...
-        setAccessible(m, true);
+        // reject non-public premain or agentmain method
+        if ((m.getModifiers() & java.lang.reflect.Modifier.PUBLIC) == 0) {
```

Comment:
An alternative way: `if (Modifier.isPublic(m.getModifiers()))`

Similar can be applied for checking if the javaagentClass is public.